src/waveform/renderers/qopengl/colordata.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <QVector3D>
+#include <QVector>
+
+namespace qopengl {
+class ColorData;
+}
+
+class qopengl::ColorData {
+    QVector<QVector3D> mData;
+
+  public:
+    void addForRectangle(float r, float g, float b) {
+        mData.push_back({r, g, b});
+        mData.push_back({r, g, b});
+        mData.push_back({r, g, b});
+        mData.push_back({r, g, b});
+        mData.push_back({r, g, b});
+        mData.push_back({r, g, b});
+    }
```

Comment:
Yes, you guessed correct.

I am not quite sure would you mean with "implicitly coupled to the topology of that rectangle." :-) But do you have some particular other approach in mind? Or is this something that could simply be addressed with a more explicit function name?

Reply:
I am waiting for some clarification / suggestion on how to improve this. But I don't think it should block this PR, it's something that can be done later. It works.

Reply:
Right, so to elaborate: The operations here are tied to how opengl is actually interpreting the underlying mesh. You could use this method, but when you actually draw the mesh, it could be specified as `GL_LINE` or something other than `GL_TRIANGLE*`. The result will obviously look nothing alike. So the issue I have with this is that it tries to abstract the vertices into primitives, but that abstraction is incomplete because vertices have more than just position information (such as color). Color and position information are related to each other via their index in the array but this abstraction removes that. So its now very easy to create vertices without color (or rather vertices which are matched with the wrong color). So for this abstraction to work properly, it would need to work by taking all attributes of a vertex at once. In the end, it really boils down to the classic "struct of arrays vs array of structs" problem. 

So no, it doesn't block this PR, but IMO its not a huge improvement either because of its leaky-ness.

Reply:
Ok, I see what you mean, and I have some ideas on how to address this. But I feel that we are piling reviews upon reviews, so let's just stick with this solution for now and in the future do a refactoring on top of this.

Reply:
Yes, I agree.